### PR TITLE
Single Version Query Logic

### PR DESCRIPTION
**Pull request by NobleMathews**
Implements single version querying
Adds support for parsing non-specific requirements
Modified semver parsing with assumptions made for version not compliant with semver
Fixes Postgres DB return type and integration with other responses

### CODE REVIEW

File: dependencies/helper.py [hunk 1]
```diff
@@ -372,3 +377,92 @@ def php_versions(api_response: requests.Response, queries: dict) -> list:
...
+def is_non_specific(req: str) -> bool:
+    """
+    Checks if requirement string has any character that makes it non-specific
+    :param req: requirement string
+    """
+    non_sp_list = [">", "<", "~", "^", "*", "!", ",", "latest"]
+    return any(i in req for i in non_sp_list)
+
+
+def resolve_version(vers: List[str], reqs=None) -> str:
```
Comment (tushar-deepsource) [inline dependencies/helper.py:391]: This will have to be refactored into self-sufficient documented functions

**Review by tushar-deepsource — CHANGES_REQUESTED** [review]

File: dependencies/py/py_helper.py [hunk 2]
```diff
@@ -27,10 +27,5 @@ def handle_requirements_txt(req_file_data: str) -> Result:
         if not ir.specs:
             res["pkg_dep"].append(str(ir.key))
         else:
-            for spec in ir.specs:
-                if "=" in str(spec[0]):
-                    res["pkg_dep"].append(str(ir.key) + ";" + str(spec[1]))
-                else:
-                    # ! Handle each case properly
-                    res["pkg_dep"].append(str(ir.key))
+            res["pkg_dep"].append(str(ir.key) + "|" + str(ir.specs))
```
[the remark this answers is not in the document]
Comment (NobleMathews) [reply dependencies/py/py_helper.py:30]: Will do it in another PR, need to modify quite a bit

File: dependencies/helper.py [hunk 3]
```diff
@@ -372,3 +377,142 @@ def php_versions(api_response: requests.Response, queries: dict) -> list:
...
+            # Inequality requirements
+            elif sym == "!=":
+                compatible_vers = list(filter(lambda x: x != ver, vers))
+            elif ">" in sym:
+                compatible_vers = list(
+                    filter(find_newer_versions(ver, include_ver="=" in sym), vers)
+                )
+            elif "<" in sym:
+                compatible_vers = list(
+                    filter(find_older_versions(ver, include_ver="=" in sym), vers)
```
Comment (tushar-deepsource) [inline dependencies/helper.py:428]: The logic here is a bit weird to me, and I find the code of `find_older_versions` a bit hard to follow. And the `x in sym` checks I don't like, for example.

Can we do something like this instead?

```python
elif sym in ('>', '>=', '<', '<='):
    parsed_versions = [VersionInfo.parse(version) for version in vers]
    version = VersionInfo.parse(ver)
    if sym == '>':
        compatible_vers = [str(x) for x in parsed_versions x > version]
    if sym == '>=':
        compatible_vers = [str(x) for x in parsed_versions x >= version]
    if sym == '<':
        compatible_vers = [str(x) for x in parsed_versions x < version]
    if sym == '<=':
        compatible_vers = [str(x) for x in parsed_versions x <= version]
```

**Review by tushar-deepsource — CHANGES_REQUESTED** [review]

File: dependencies/helper.py [hunk 4]
```diff
@@ -372,3 +377,121 @@ def php_versions(api_response: requests.Response, queries: dict) -> list:
...
+    Compatible versions based on specified major, minor, and patch version
+    ~1.2.3	>=1.2.3 <1.3.0
+    ~1.2	>=1.2.0 <1.3.0
+    ~1	    >=1.0.0 <2.0.0
+    """
+    dot_count = ver.count(".")
+    svi = semver.VersionInfo.parse(ver)
+    if dot_count in (1, 0):
+        vers = [x for x in vers if semver.VersionInfo.parse(x).major == svi.major]
+    elif dot_count >= 2:
```
Comment (tushar-deepsource) [inline dependencies/helper.py:457]:
```suggestion
    if dot_count <= 1:
        vers = [x for x in vers if semver.VersionInfo.parse(x).major == svi.major]
    else:
```

**Review by tushar-deepsource — DISMISSED** [review]

**Review by tushar-deepsource — APPROVED** [review]
